src/Base/SmallMatrix.H

```diff
@@ -0,0 +1,319 @@
+/* Copyright 2021-2025 The AMReX Community
+ *
+ * Authors: Axel Huebl
+ * License: BSD-3-Clause-LBNL
+ */
+#pragma once
+
+#include "pyAMReX.H"
+
+#include <AMReX_SmallMatrix.H>
+
+#include <complex>
+#include <cstdint>
+#include <iterator>
+#include <sstream>
+#include <type_traits>
+#include <vector>
+
+
+namespace
+{
+    // helper type traits
+    template <typename T>
+    struct get_value_type { using value_type = T; };
+    template <typename T>
+    struct get_value_type<std::complex<T>> { using value_type = T; };
+    template <typename T>
+    using get_value_type_t = typename get_value_type<T>::value_type;
+
+    // helper to check if Array4<T> is of constant value type T
+    template <typename T>
+    constexpr bool is_not_const ()
+    {
+        return std::is_same_v<
+            std::remove_cv_t<
+                T
+            >,
+            T
+        > &&
+        std::is_same_v<
+            std::remove_cv_t<
+                get_value_type_t<T>
+            >,
+            get_value_type_t<T>
+        >;
+    }
+}
+
+namespace pyAMReX
+{
+    using namespace amrex;
+
+    /** CPU: __array_interface__ v3
+     *
+     * https://numpy.org/doc/stable/reference/arrays.interface.html
+     */
+    template<
+        class T,
+        int NRows,
+        int NCols,
+        Order ORDER = Order::F,
+        int StartIndex = 0
+    >
+    py::dict
+    array_interface (SmallMatrix<T, NRows, NCols, ORDER, StartIndex> const & m)
+    {
+        auto d = py::dict();
+        // provide C index order for shape and strides
+        auto shape = m.ordering == Order::F ? py::make_tuple(
+            py::ssize_t(NRows),
+            py::ssize_t(NCols)  // fastest varying index
+        ) : py::make_tuple(
+            py::ssize_t(NCols),
+            py::ssize_t(NRows)  // fastest varying index
+        );
+        // buffer protocol strides are in bytes
+        auto const strides = m.ordering == Order::F ? py::make_tuple(
+            py::ssize_t(sizeof(T) * NCols),
+            py::ssize_t(sizeof(T))  // fastest varying index
+        ) : py::make_tuple(
+            py::ssize_t(sizeof(T) * NRows),
+            py::ssize_t(sizeof(T))  // fastest varying index
+        );
+        bool const read_only = false;  // note: we could decide on is_not_const,
+                                       // but many libs, e.g. PyTorch, do not
+                                       // support read-only and will raise
+                                       // warnings, casting to read-write
+        d["data"] = py::make_tuple(std::intptr_t(&m.template get<0>()), read_only);
```

Comment:
@WeiqunZhang should `.get` honor the StartIndex in AMReX?
https://github.com/WeiqunZhang/amrex/blob/3fcfcce263bdfdb80297336284b3e9c07c85060d/Src/Base/AMReX_SmallMatrix.H#L392-L398